arch/risc-v/src/bl602/bl602_os_hal.c

```diff
@@ -0,0 +1,1035 @@
+ * Input Parameters:
+ *
+ * Returned Value:
+ *
+ ****************************************************************************/
+
+void bl_os_irq_attach(int32_t n, void *f, void *arg)
+{
+  int ret;
+  struct irq_adpt *adapter;
+
+  wlinfo("INFO: n=%ld f=%p arg=%p\n", n, f, arg);
+
+  adapter = (struct irq_adpt *)kmm_malloc(sizeof(struct irq_adpt));
+
+  if (!adapter)
+    {
+      DEBUGASSERT(0);
+    }
+
+  adapter->func = f;
+  adapter->arg  = arg;
+
+  ret = irq_attach(n, bl_os_isr_adpt_cb, (void *)adapter);
+
+  if (ret != OK)
+    {
+      DEBUGASSERT(0);
+    }
+}
+
+/****************************************************************************
+ * Name: bl_os_mutex_create
+ *
+ * Description:
+ *   Create mutex
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   Mutex data pointer
+ *
+ ****************************************************************************/
+
+void *bl_os_mutex_create(void)
+{
+  int ret;
+  pthread_mutex_t *mutex;
+  int tmp;
+
+  tmp = sizeof(pthread_mutex_t);
+  mutex = (pthread_mutex_t *)kmm_malloc(tmp);
+  if (!mutex)
+    {
+      wlerr("ERROR: Failed to alloc %d memory\n", tmp);
+      return NULL;
+    }
+
+  ret = pthread_mutex_init(mutex, NULL);
+  if (ret)
+    {
+      wlerr("ERROR: Failed to initialize mutex error=%d\n", ret);
+      kmm_free(mutex);
+      return NULL;
+    }
+
+  return mutex;
+}
+
+/****************************************************************************
+ * Name: bl_os_mutex_delete
+ *
+ * Description:
+ *   Delete mutex
+ *
+ * Input Parameters:
+ *   mutex_data - mutex data pointer
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void bl_os_mutex_delete(void *mutex_data)
+{
+  pthread_mutex_t *mutex = (pthread_mutex_t *)mutex_data;
+
+  pthread_mutex_destroy(mutex);
+  kmm_free(mutex);
+}
+
+/****************************************************************************
+ * Name: bl_os_mutex_lock
+ *
+ * Description:
+ *   Lock mutex
+ *
+ * Input Parameters:
+ *   mutex_data - mutex data pointer
+ *
+ * Returned Value:
+ *   True if success or false if fail
+ *
+ ****************************************************************************/
+
+int32_t bl_os_mutex_lock(void *mutex_data)
+{
+  int ret;
+  pthread_mutex_t *mutex = (pthread_mutex_t *)mutex_data;
+
+  ret = pthread_mutex_lock(mutex);
+  if (ret)
+    {
+      wlerr("ERROR: Failed to lock mutex error=%d\n", ret);
+    }
+
+  return bl_os_errno_trans(ret);
+}
+
+/****************************************************************************
+ * Name: bl_os_mutex_unlock
+ *
+ * Description:
+ *   Lock mutex
+ *
+ * Input Parameters:
+ *   mutex_data - mutex data pointer
+ *
+ * Returned Value:
+ *   True if success or false if fail
+ *
+ ****************************************************************************/
+
+int32_t bl_os_mutex_unlock(void *mutex_data)
+{
+  int ret;
+  pthread_mutex_t *mutex = (pthread_mutex_t *)mutex_data;
```

Comment:
switch to semaphore

Reply:
Why is it not recommended to use pthread？

Reply:
pthread API is only for userspace, all kernel component should use semaphore.